doc/neps/nep-0034.rst

```diff
@@ -0,0 +1,131 @@
+confusing, since it is rarely what is desired. Changing the automatic dtype
+detection to never return ``object`` for ragged nested sequences (defined as a
+recursive sequence of sequences, where not all the sequences on the same
+level have the same length) will force users who actually wish to create
+``object`` arrays to specify that explicitly. Note that lists, tuples, and
+nd.arrays are all sequences. See for instance `issue 5303`_.
+
+Usage and Impact
+----------------
+
+After this change, array creation with ragged nested sequences must explicitly
+define a dtype:
+
+    >>> np.array([[1, 2], [1]])
+    ValueError: cannot guess the desired dtype from the input
+
+    >>> np.array([[1, 2], [1]], dtype=object)
+    # succeeds, with no change from current behaviour
+
+Detailed description
+--------------------
+
+To explicitly set the shape of the object array, since it is sometimes hard to
+determine what shape is desired, one could use:
+
+    >>> arr = np.empty(correct_shape, dtype=object)
+    >>> arr[...] = values
+
+We will also reject mixed sequences of non-sequence and sequence, for instance
+all of these will be rejected:
+
+    >>> arr = np.array([np.arange(10), [10]])
+    >>> arr = np.array([[range(3), range(3), range(3)], [range(3), 0, 0]])
+
+Related Work
+------------
+
+`PR 14341`_ tried to raise an error when ragged nested sequences were specified
+with a numeric dtype ``np.array, [[1], [2, 3]], dtype=int)`` but failed due to
+false-positives, for instance ``np.array([1, np.array([5])], dtype=int)``.
+
+.. _`PR 14341`: https://github.com/numpy/numpy/pull/14341
+
+Implementation
+--------------
+
+The code to be changed is inside ``PyArray_GetArrayParamsFromObject`` and the
+internal ``discover_dimentions`` function. See `PR 14794`_.
+
+Backward compatibility
+----------------------
+
+Anyone depending on ragged nested sequences creating object arrays will need to
+modify their code. There will be a deprecation period during which the current
+behaviour will emit a ``DeprecationWarning``. There is an open question whether
+the ``assert_equal`` family of functions should be changed or users be forced
+to change code like
+
+```
+np.assert_equal(a, [[1, 2], 3])
+```
+
+to 
+
+```
+np.assert_equal(a, np.array([[1, 2], 3], dtype=object)
+```
+
+Alternatives
+------------
+
+- We could continue with the current situation.
+
+- It was also suggested to add a kwarg ``depth`` to array creation, or perhaps
+  to add another array creation API function ``ragged_array_object``. The goal
+  was to eliminate the ambiguity in creating an object array from ``array([[1,
+  2], [1]], dtype=object)``: should the returned array have a shape of
+  ``(1,)``, or ``(2,)``? This NEP does not deal with that issue, and only
+  deprecates the use of ``array`` with no ``dtype=object`` for ragged nested
+  sequences. Users of ragged nested sequences may face another deprecation
+  cycle in the future.
```

Comment:
I'd add something like: _Rationale: we expect that there are very few users who intend to use ragged arrays like that, this was never intended as a use case of NumPy arrays. Users are likely better off with another library or just using list of lists._

Reply:
The phrasing I was pushing for earlier was more along the lines of "this isn't a big enough problem to be worth bringing into the scope of this NEP", rather than "this isn't something in scope for numpy". I'd still consider attempting to solve this in some future NEP. The rationale in my mind was simply "this is a different problem that can be solved later, and is lower value than the contents of this NEP".

Reply:
That phrasing works for me too.

> I'd still consider attempting to solve this in some future NEP

I would suggest that we have many more interesting/important things to do. But yeah, we've never had the discussion so let's just postpone it rather than discussing it now.